server/controler/register_test.go

```diff
@@ -0,0 +1,102 @@
+	"github.com/rungokarol/facilEspanol/model"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+)
+
+type RegisterTestSuite struct {
+	suite.Suite
+	env       *Env
+	storeMock *mocks.IDataStore
+	rr        *httptest.ResponseRecorder
+	handler   http.Handler
+}
+
+func (suite *RegisterTestSuite) SetupTest() {
+	suite.storeMock = &mocks.IDataStore{}
+	suite.env = CreateEnv(suite.storeMock)
+	suite.rr = httptest.NewRecorder()
+	suite.handler = http.HandlerFunc(suite.env.Register)
+}
+
+func TestRegister(t *testing.T) {
+	suite.Run(t, new(RegisterTestSuite))
+}
+
+func (suite *RegisterTestSuite) TestRejectWithNotPostMethod() {
+	req, err := http.NewRequest("GET", "/user/register", nil)
+	assert.Nil(suite.T(), err)
+
+	suite.handler.ServeHTTP(suite.rr, req)
+	assert.Equal(suite.T(), http.StatusMethodNotAllowed, suite.rr.Code)
+}
+
+func (suite *RegisterTestSuite) TestRejectWhenBodyIsNotJson() {
+	notJsonBody := ioutil.NopCloser(bytes.NewBufferString("Hello World"))
+	req, err := http.NewRequest("POST", "/user/register", notJsonBody)
+	assert.Nil(suite.T(), err)
+
+	suite.handler.ServeHTTP(suite.rr, req)
+	assert.Equal(suite.T(), http.StatusBadRequest, suite.rr.Code)
+}
+
+func (suite *RegisterTestSuite) TestRejectIfUsernameTooShort() {
+	jsonBody, err := json.Marshal(registerReq{Username: "ja", Password: "bar"})
+	assert.Nil(suite.T(), err)
+	req, err := http.NewRequest("POST", "/user/register", bytes.NewBuffer(jsonBody))
+	assert.Nil(suite.T(), err)
+
+	suite.handler.ServeHTTP(suite.rr, req)
+	assert.Equal(suite.T(), http.StatusBadRequest, suite.rr.Code)
+}
+
+func (suite *RegisterTestSuite) TestRejectIfPasswordTooShort() {
+	jsonBody, err := json.Marshal(registerReq{Username: "foo", Password: "ja"})
+	assert.Nil(suite.T(), err)
+	req, err := http.NewRequest("POST", "/user/register", bytes.NewBuffer(jsonBody))
+	assert.Nil(suite.T(), err)
+
+	suite.handler.ServeHTTP(suite.rr, req)
+	assert.Equal(suite.T(), http.StatusBadRequest, suite.rr.Code)
+}
+
+func (suite *RegisterTestSuite) TestRejectIfUserAlreadyExists() {
+	username := "foo"
+	jsonBody, err := json.Marshal(registerReq{Username: username, Password: "bar"})
+	assert.Nil(suite.T(), err)
+	req, err := http.NewRequest("POST", "/user/register", bytes.NewBuffer(jsonBody))
+	assert.Nil(suite.T(), err)
+
+	suite.storeMock.On("IsUserPresent", username).Return(true, nil)
+
+	suite.handler.ServeHTTP(suite.rr, req)
+	assert.Equal(suite.T(), http.StatusBadRequest, suite.rr.Code)
+}
+
+func (suite *RegisterTestSuite) TestAcceptRegistration() {
+	username := "foo"
+	jsonBody, err := json.Marshal(registerReq{Username: username, Password: "bar"})
+	assert.Nil(suite.T(), err)
+	req, err := http.NewRequest("POST", "/user/register", bytes.NewBuffer(jsonBody))
+	assert.Nil(suite.T(), err)
+
+	suite.storeMock.
+		On("IsUserPresent", username).Return(false, nil).
+		On("CreateUser", mock.Anything).Return(nil);
```

Comment:
can we match the username here and delete line nr 100?

Reply:
I'm afraid it's not possible @rungokarol. Setting mocks and asserting if they have been called with expected arguments are two separate concerns. What I could make is to replace `mock.Anything` with more strict matcher, I've used for the assert below.

Reply:
It is not necessary because we're matching this below ;)